Sources/SourceKitLSPAPI/BuildDescription.swift

```diff
@@ -142,4 +142,26 @@ public struct BuildDescription {
             getBuildTarget(for: $0, in: modulesGraph)
         }
     }
+
+    /// Returns `true` if the file at the given path might influence build settings for a `swiftc` or `clang` invocation
+    /// generated by SwiftPM.
+    public func fileAffectsSwiftOrClangBuildSettings(_ url: URL) -> Bool {
+        guard let filePath = try? AbsolutePath(validating: url.path) else {
+            return false
+        }
+
+        for input in buildPlan.inputs {
```

Comment:
Could we unify this filtering with inputs somehow to avoid having to build inputs just to want them again to find one relevant location?

Reply:
I don’t think it’s worth optimizing for that:
- I don’t think that building up the inputs is all that expensive and will be any performance impact in practice. The only variables here are:
  - Number of root packages, which will be 1 in almost all cases
  - Number of source files in the root package, which is usually small-ish, definitely doesn’t go quadratic or something like that
  - Number of prebuilt output directories (I also expect this to be <10)
- The code would become quite a bit more complicated if we want to share this method for `LLBuildManifest` and SourceKit-LSP

Reply:
I've recently been shown a project with a large number of targets which was already a problem. I'm not sure how many times would `fileAffectsSwiftOrClangBuildSettings` this be called in SourceKit-LSP but it seems like that would be a problem for that project at least. If we could at least cache the `inputs` that would make a difference.

Reply:
Oh, interesting. Looks like my performance judging wasn’t correct then.

Caching is the right solution here then. Otherwise we’d still get the worst-case behavior when a file is passed to `fileAffectSwiftOrClangBuildSettings` that’s not part of the inputs.